pkg/bootkube/bootkube.go

```diff
@@ -31,141 +23,65 @@ var requiredPods = []string{
-		"--tls-cert-file=" + filepath.Join(config.AssetDir, asset.AssetPathAPIServerCert),
-		"--kubelet-client-key=" + filepath.Join(config.AssetDir, asset.AssetPathAPIServerKey),
-		"--kubelet-client-certificate=" + filepath.Join(config.AssetDir, asset.AssetPathAPIServerCert),
-		"--client-ca-file=" + filepath.Join(config.AssetDir, asset.AssetPathCACert),
-		"--authorization-mode=RBAC",
-		"--etcd-servers=" + etcdServers,
-		"--service-cluster-ip-range=" + serviceCIDR,
-		"--service-account-key-file=" + filepath.Join(config.AssetDir, asset.AssetPathServiceAccountPubKey),
-		"--admission-control=NamespaceLifecycle,ServiceAccount",
-		"--runtime-config=api/all=true",
-		"--storage-backend=etcd3",
-	}, nil
-}
 
-func makeControllerManagerFlags(config Config) ([]string, error) {
-	podCIDR, err := detectPodCIDR(config.AssetDir)
-	if err != nil {
-		return []string{}, err
+	if err = CreateAssets(filepath.Join(b.assetDir, asset.AssetPathManifests), assetTimeout); err != nil {
+		return err
 	}
-	return []string{
-		"--master=" + insecureAPIAddr,
-		"--service-account-private-key-file=" + filepath.Join(config.AssetDir, asset.AssetPathServiceAccountPrivKey),
-		"--root-ca-file=" + filepath.Join(config.AssetDir, asset.AssetPathCACert),
-		"--allocate-node-cidrs=true",
-		"--cluster-cidr=" + podCIDR,
-		"--configure-cloud-routes=false",
-		"--leader-elect=true",
-	}, nil
-}
 
-func (b *bootkube) Run() error {
-	UserOutput("Running temporary bootstrap control plane...\n")
-
-	errch := make(chan error)
-	go func() { errch <- apiapp.Run(b.apiServer) }()
-	go func() { errch <- cmapp.Run(b.controller) }()
-	go func() { errch <- schedapp.Run(b.scheduler) }()
-	go func() {
-		if err := CreateAssets(filepath.Join(b.assetDir, asset.AssetPathManifests), assetTimeout); err != nil {
-			errch <- err
+	if b.selfHostedEtcd {
+		requiredPods = append(requiredPods, "etcd-operator")
+	}
+
+	if err = WaitUntilPodsRunning(requiredPods, assetTimeout); err != nil {
+		return err
+	}
+
+	if b.selfHostedEtcd {
+		UserOutput("Migrating to self-hosted etcd cluster...\n")
+		var etcdServiceIP string
+		etcdServiceIP, err = detectEtcdIP(b.assetDir)
+		if err != nil {
+			return err
 		}
-	}()
-	go func() {
-		if b.selfHostedEtcd {
-			requiredPods = append(requiredPods, "etcd-operator")
-			etcdServiceIP, err := detectEtcdIP(b.assetDir)
-			if err != nil {
-				errch <- err
-				return
-			}
-			if err := WaitUntilPodsRunning(requiredPods, assetTimeout); err != nil {
-				errch <- err
-				return
-			}
-			errch <- etcdutil.Migrate(etcdServiceIP)
-		} else {
-			errch <- WaitUntilPodsRunning(requiredPods, assetTimeout)
+		if err = etcdutil.Migrate(insecureAPIAddr, etcdServiceIP); err != nil {
+			return err
 		}
-	}()
-	// If any of the bootkube services exit, it means it is unrecoverable and we should exit.
-	err := <-errch
-	if err != nil {
-		UserOutput("Error: %v\n", err)
 	}
-	return err
+
+	if err = CleanupBootstrapControlPlane(b.assetDir, b.podManifestPath); err != nil {
```

Comment:
Can this be moved to the top of this function but using `defer` This way assets get cleaned up even if bootkube has an error before reaching this function.